### PR TITLE
DATACMNS-1035 - Moved CustomConversions to Spring Data Commons.

### PR DESCRIPTION
**Pull request by olivergierke**
Extracted CustomConversions — whose code has largely been duplicated between the MongoDB, Couchbase and Cassandra modules — into Spring Data Commons. Store-specific extensions can now be contributed via a StoreConversions value type that carries both, store-specific default converters as well as a store-specific SimpleTypeHolder to augment the default list of simple types.

Removed SimpleTypeHolders public default constructor in favour of a protected one and a static DEFAULT instance for plain references.

### CODE REVIEW

File: src/main/java/org/springframework/data/convert/CustomConversions.java [hunk 1]
```diff
@@ -0,0 +1,489 @@
...
+		 * Creates a new {@link StoreConversions} for the given store-specific {@link SimpleTypeHolder} and the given
+		 * converters.
+		 * 
+		 * @param storeTypeHolder must not be {@literal null}.
+		 * @param converters must not be {@literal null}.
+		 * @return
+		 */
+		public static StoreConversions of(SimpleTypeHolder storeTypeHolder, Object... converters) {
+
+			Assert.notNull(storeTypeHolder, "SimpleTypeHolder must not be null!");
```
Comment (mp911de) [inline src/main/java/org/springframework/data/convert/CustomConversions.java]: These two overloads make it quite easy to use by accident the wrong method. Should check here that `converters` isn't a single `Collection` element?

File: src/main/java/org/springframework/data/convert/CustomConversions.java [hunk 2]
```diff
@@ -0,0 +1,489 @@
...
+		this.rawWriteTargetTypes = new ConcurrentHashMap<>();
+
+		List<Object> toRegister = new ArrayList<Object>();
+
+		// Add user provided converters to make sure they can override the defaults
+		toRegister.addAll(converters);
+		toRegister.addAll(JodaTimeConverters.getConvertersToRegister());
+		toRegister.addAll(Jsr310Converters.getConvertersToRegister());
+		toRegister.addAll(ThreeTenBackPortConverters.getConvertersToRegister());
+		toRegister.addAll(storeConversions.getStoreConverters());
```
Comment (mp911de) [inline src/main/java/org/springframework/data/convert/CustomConversions.java]: Store converters should be probably registered after `converters` and before `JodaTimeConverters` to retain registration order. MongoDB and Cassandra modules register their converters after `converters` and before `JodaTimeConverters`.

File: src/main/java/org/springframework/data/convert/CustomConversions.java [hunk 3]
```diff
@@ -0,0 +1,489 @@
...
+		Assert.notNull(converters, "List of converters must not be null!");
+
+		this.readingPairs = new LinkedHashSet<>();
+		this.writingPairs = new LinkedHashSet<>();
+		this.customSimpleTypes = new HashSet<>();
+		this.customReadTargetTypes = new ConcurrentHashMap<>();
+		this.customWriteTargetTypes = new ConcurrentHashMap<>();
+		this.rawWriteTargetTypes = new ConcurrentHashMap<>();
+
+		List<Object> toRegister = new ArrayList<Object>();
```
Comment (gregturn) [inline src/main/java/org/springframework/data/convert/CustomConversions.java]: At least can say `new ArrayList<>()`

File: src/main/java/org/springframework/data/convert/CustomConversions.java [hunk 4]
```diff
@@ -0,0 +1,489 @@
...
+
+		// Add user provided converters to make sure they can override the defaults
+		toRegister.addAll(converters);
+		toRegister.addAll(JodaTimeConverters.getConvertersToRegister());
```
Comment (gregturn) [inline src/main/java/org/springframework/data/convert/CustomConversions.java]: +1

File: src/main/java/org/springframework/data/convert/CustomConversions.java [hunk 5]
```diff
@@ -0,0 +1,489 @@
...
+				added = true;
+			}
+
+			if (it instanceof GenericConverter) {
+				conversionService.addConverter(GenericConverter.class.cast(it));
+				added = true;
+			}
+
+			if (!added) {
+				throw new IllegalArgumentException(
```
Comment (gregturn) [inline src/main/java/org/springframework/data/convert/CustomConversions.java]: I'd favor logging each registration.  Then exception can say "couldn't find match". Also avoid tracking state of who was registered.

File: src/main/java/org/springframework/data/convert/CustomConversions.java (hunk 3 above)
Comment (olivergierke) [inline src/main/java/org/springframework/data/convert/CustomConversions.java]: Is there really a benefit of calling `….stream()` on each of the `List`s? The constructor invocation doesn't need an explicit generic parameter, right.

File: src/main/java/org/springframework/data/convert/CustomConversions.java (hunk 5 above)
Comment (olivergierke) [inline src/main/java/org/springframework/data/convert/CustomConversions.java]: Exposing the offending instance is a good idea. Is there added value in logging the converters registered? If so, shouldn't this be functionality `GenericConversionService`?

File: src/main/java/org/springframework/data/convert/CustomConversions.java (hunk 1 above)
Comment (olivergierke) [inline src/main/java/org/springframework/data/convert/CustomConversions.java]: How so? You'd explicitly need to call `StoreConversions.of(new Object[] { collection })` to run into this problem, right?

File: src/main/java/org/springframework/data/mapping/model/SimpleTypeHolder.java [hunk 6]
```diff
@@ -32,39 +32,43 @@
...
-		DEFAULTS.add(Float.class);
-		DEFAULTS.add(Double.class);
-		DEFAULTS.add(Character.class);
-		DEFAULTS.add(String.class);
-		DEFAULTS.add(Date.class);
-		DEFAULTS.add(Locale.class);
-		DEFAULTS.add(Class.class);
-		DEFAULTS.add(Enum.class);
-	}
+	private static final Set<Class<?>> DEFAULTS = new HashSet<Class<?>>() {
```
Comment (olivergierke) [inline src/main/java/org/springframework/data/mapping/model/SimpleTypeHolder.java]: It's the only way you can have a static `DEFAULT` instance see below *and* get the `DEFAULTS` `HashSet` populated already, which is used in the constructor. If `DEFAULTS` is set up in a static initializer block, it's still empty when the constructor is called to create the `DEFAULT` instance.

File: src/main/java/org/springframework/data/convert/CustomConversions.java (hunk 4 above)
Comment (olivergierke) [inline src/main/java/org/springframework/data/convert/CustomConversions.java]: I didn't bother to again extract the converters here into a static `List` as there should be only one production instance of `CustomConversions` in place anyway. We'd basically keep yet another `List` around and introduce a layer of indirection. That said, we can definitely extract that list, no problem.

**Comment by olivergierke** [comment]
From the mixture of feedback I conclude the following steps.

1. Switch to `StoreConversions.of(…, Collection<?>)` to avoid the usage of the potentially wrong overload.
2. Create a static list for the default converters we register in `CustomConversions`.
3. Stick to `addAll(…)` for final registration as we then align the abstraction level of the registration steps: user converters, store converters, default ones.
4. As 3 indicates, fix the order of registration to register the default converters last.
5. Improve exception message on converter registration.

**Pull request closed without merging**